File: dsa/dp/dogecoin/solutions/dogecoin.js
const solve = () => {
    const gets = require('readline-sync').question;
    const print = console.log;
    const quit = () => {};

    const [rows, cols] = gets().split(' ').map(Number);

    const coins = parseInt(gets());

    let field = Array.from({length: rows})
        .map(() => Array.from({length: cols})
            .map(() => 0)
        );

    for(let i = 0; i < coins; i += 1) {
        [r, c] = gets().split(' ').map(Number);
        field[r][c] += 1;
    }

    for(let i = 1; i < rows; i += 1) {
        field[i][0] += field[i-1][0];
    }

    for(let i = 1; i < cols; i += 1) {
        field[0][i] += field[0][i-1];
    }

    for(let row = 1; row < rows; row += 1) {
        for(let col = 1; col < cols; col += 1) {
            field[row][col] += Math.max(field[row-1][col], field[row][col-1]);
        }
    }

    print(field[rows-1][cols-1])
    quit();
    return;
};

solve();

File: easy/14dp/02dogecoin/solutions/dogecoin.swift
let nm = readLine()!
    .characters.split() {$0 == " "}
    .map(String.init)
    .map() {Int($0)!};

let rows = nm[0]
let cols = nm[1]

var field: [[Int]] = []

for r in 0..<rows {
    field.append([]);
    for _ in 0..<cols {
        field[r].append(0)
    } 
}

let coins = Int(readLine()!)!

for _ in 0..<coins {
    let rc = readLine()!
        .characters.split() {$0 == " "}
        .map(String.init)
        .map() {Int($0)!};
    field[rc[0]][rc[1]] += 1
}

for row in 1..<rows {
    field[row][0] += field[row-1][0]
}

for col in 1..<cols {
    field[0][col] += field[0][col-1]
}

for row in 1..<rows {
    for col in 1..<cols {
        let a = field[row - 1][col]
        let b = field[row][col - 1]
        if(a < b) {
            field[row][col] += b
        } else {
            field[row][col] += a
        }
    }
}

print(field[rows-1][cols-1])